### PR TITLE
Config merge

### LINKED ISSUE #12: Merging in user configuration
Would it be in scope for this project to merge existing Pulumi config with the localstack config? The functionality would be such that if a user provided a `--config-file` argument to `pulumilocal`, it would merge that with the `Pulumi.localstack.yaml` file.

I'm pretty new to Pulumi so wonder if I'm overlooking another way to do this.

### PR DESCRIPTION
**Pull request by lakkeger**
# Motivation
Config merging and respecting the original stack config is not something currently well supported in `pulumi-local`. To solve this issue, reported on this [issue](#12) too, we'll introduce 3 different config strategies:
- **overwrite strategy (current, default):** this config option simply updates/overwrites the current config with the necessary changes we have to apply for LocalStack
- **override (with dry-run):** this config option temporarily creates a config file by merging the original stacks config with our recommended options and running pulumi commands with it. By setting dry-run, `pulumi-local` creates the config file then terminates without running the command.
- **separation:** copying the currently selected stack's config and creates a new separate stack (by default called localstack, however this is configurable with `DEFAULT_STACK_NAME` env variable)

If any of the config files we'd like to create exists we'll prompt for user input.

## Additional options and features
- `DRY_RUN`: in cases when makes sense creates the config file but terminates after without modifying it, if it finds an already existing one it prompts the user.
- `--non-interactive`/`NON_INTERACTIVE`: the usage of the non-interactive flag is now allowed to suppress any `pulumi-local` prompts for user input.
-  `LOCAL_STACK_NAME`: the name of the LocalStack stack/config file we create, depending on the strategy
- deactivate `aws:accessKey` and `aws:secretKey` options

# Changes
- [x] add user prompt for existing stack configs
- [x] add logic for strategy selection
- [x] add merge strategy, by taking leverage of `--config-file` option of pulumi
- [x] add separate stack strategy by copying the currently selected stack's config and then change to it
- [x] deactivate AWS credentials in the config file
- [x] create tests for strategies

### CODE REVIEW

File: bin/pulumilocal [hunk 1]
```diff
@@ -23,17 +24,62 @@ if os.path.isdir(os.path.join(PARENT_FOLDER, ".venv")):
 
 # define global constants
 TRUE_STRINGS = ["1", "true"]
+CONFIG_STRATEGIES = ("overwrite", "override", "separation")
+DEFAULT_CONFIG_STRATEGY = "overwrite"
+LOCAL_STACK_NAME = os.environ.get("LOCAL_STACK_NAME") or "localstack"
```
Comment (simonrw) [inline bin/pulumilocal]: nit: this is a little confusing since it's very similar to `LOCALSTACK`. I get that this name represents the stack name that you use to run locally, but I wonder if the confusion is too much.

File: bin/pulumilocal [hunk 2]
```diff
@@ -23,17 +24,62 @@ if os.path.isdir(os.path.join(PARENT_FOLDER, ".venv")):
...
+CONFIG_STRATEGY = os.environ.get("CONFIG_STRATEGY") if os.environ.get("CONFIG_STRATEGY") in CONFIG_STRATEGIES else DEFAULT_CONFIG_STRATEGY
```
Comment (simonrw) [inline bin/pulumilocal:30]: We should probably warn the user if they have specified an invalid strategy. I.e. print an error message and exit with bad exit code?

File: bin/pulumilocal [hunk 3]
```diff
@@ -23,17 +24,62 @@ if os.path.isdir(os.path.join(PARENT_FOLDER, ".venv")):
...
 PULUMI_CMD = os.environ.get("PULUMI_CMD") or "pulumi"
 AWS_ENDPOINT_URL = os.environ.get("AWS_ENDPOINT_URL")
 LOCALSTACK_HOSTNAME = urlparse(os.environ.get("AWS_ENDPOINT_URL")).hostname or os.environ.get("LOCALSTACK_HOSTNAME") or "localhost"
 EDGE_PORT = int(urlparse(os.environ.get("AWS_ENDPOINT_URL")).port or os.environ.get("EDGE_PORT") or 4566)
 USE_SSL = str(os.environ.get("USE_SSL")).strip().lower() in TRUE_STRINGS
+DRY_RUN = str(os.environ.get("DRY_RUN")).strip().lower() in TRUE_STRINGS
+NON_INTERACTIVE = str(os.environ.get("NON_INTERACTIVE")).strip().lower() in TRUE_STRINGS
```
Comment (simonrw) [inline bin/pulumilocal]: super minor: I know this is a fault with the envars above, but if `os.environ.get` returns `None` then the string that's compared with the `TRUE_STRINGS` is `None`, not `""` 😂 

```suggestion
DRY_RUN = str(os.environ.get("DRY_RUN", "")).strip().lower() in TRUE_STRINGS
NON_INTERACTIVE = str(os.environ.get("NON_INTERACTIVE", "")).strip().lower() in TRUE_STRINGS
```

File: bin/pulumilocal [hunk 4]
```diff
@@ -124,22 +193,79 @@ def set_localstack_pulumi_config(args: argparse.Namespace):
...
-    parser.add_argument("-s", "--stack", help="pulumi stack name",
+    parser.add_argument("-s", "--stack", help="The name of the stack to operate on. Defaults to the current stack",
                         required=False,
                         type=str)
-    parser.add_argument("-C", "--cwd", help="run in this directory",
+    parser.add_argument("-C", "--cwd", help="Run pulumi as if it had been started in another directory",
                         required=False,
                         type=str)
+    parser.add_argument("--config-file", help="Use the configuration values in the specified file rather than detecting the file name",
+                        required=False,
```
Comment (simonrw) [inline bin/pulumilocal:209]: Thanks for adding clarity to these help strings, these are much better.

File: bin/pulumilocal [hunk 5]
```diff
@@ -124,22 +193,79 @@ def set_localstack_pulumi_config(args: argparse.Namespace):
...
 
+    if NON_INTERACTIVE:
+        args.non_interactive = True
+
+    if not args.cwd:
+        args.cwd = "."
+
+    if os.environ.get("CONFIG_STRATEGY") not in CONFIG_STRATEGIES:
+        print(f"Config strategy: {os.environ.get('CONFIG_STRATEGY')}")
+        print(f'Config strategy is not recognised. Falling back to default ("{DEFAULT_CONFIG_STRATEGY}").')
```
Comment (simonrw) [inline bin/pulumilocal:224]: This is a duplicate from the global `CONFIG_STRATEGY`. The error message here is better, though these messages should probably print to stderr.

Also: is this a fatal error, or just a warning? If the former then we should exit the program. Perhaps that would be clearer to the user?

```suggestion
    if os.environ.get("CONFIG_STRATEGY") not in CONFIG_STRATEGIES:
        print(f"Config strategy: {os.environ.get('CONFIG_STRATEGY')}", file=sys.stderr)
        print(f'Config strategy is not recognised. Falling back to default ("{DEFAULT_CONFIG_STRATEGY}").', file=sys.stderr)
```

File: bin/pulumilocal [hunk 6]
```diff
@@ -124,22 +193,79 @@ def set_localstack_pulumi_config(args: argparse.Namespace):
...
+
     # If this is a pulumi deployment command, update the stack with LocalStack AWS config
-    if args.command in ["up", "destroy", "preview", "cancel"]:
+    if args.command in PROXIED_CMDS:
+        if not args.stack:
+            try:
+                args.stack = subprocess.run(
+                    [PULUMI_CMD, "stack", "--show-name", "--non-interactive", "-C", args.cwd],
+                    env=os.environ,
```
Comment (simonrw) [inline bin/pulumilocal]: I _think_ this is the default

File: bin/pulumilocal [hunk 7]
```diff
@@ -124,22 +193,79 @@ def set_localstack_pulumi_config(args: argparse.Namespace):
...
+                    check=True,
+                    capture_output=True
+                ).stdout.decode("utf-8").strip("\n")
+            except (subprocess.CalledProcessError) as e:
+                print(e.stderr.decode("utf-8"))
```
Comment (simonrw) [inline bin/pulumilocal]: Again, we should be printing error messages to `stderr``

```suggestion
                print(e.stderr.decode("utf-8"), file=sys.stderr)
```

File: bin/pulumilocal [hunk 8]
```diff
@@ -124,22 +193,79 @@ def set_localstack_pulumi_config(args: argparse.Namespace):
...
+                env=os.environ,
+                check=False,
+                stderr=subprocess.DEVNULL
+            )
+            override_config_file = os.path.join(os.path.dirname(config_file_path), f"Pulumi.{LOCAL_STACK_NAME}.yaml")
+            if override_config_file != config_file_path:
+                copyfile(src=config_file_path, dst=override_config_file)
+                args.config_file = override_config_file
         print("Updating this Stack with LocalStack config")
         set_localstack_pulumi_config(args)
```
Comment (simonrw) [inline bin/pulumilocal]: This set of conditions does not include the `overwrite` config strategy - is that deliberate? Is it implicitly the behaviour of lines 254-255? If so are these lines definitely compatible with being run regardless of the config strategy?

File: tests/test_apply.py [hunk 9]
```diff
@@ -75,7 +116,19 @@ def deploy_pulumi_script(script: str, version: str, select_stack: bool, select_c
         if select_cwd and not should_fail:
             cmd.extend(["--cwd", temp_dir])
         out = run(cmd, **kwargs)
-        return out[1]
+        return temp_dir
```
Comment (simonrw) [inline tests/test_apply.py:122]: I think this function should be made into a context manager, as when the temp dir is returned, `tempfile.TemporaryDirectory` will be cleaned up. This may run the cleanup at garbage collection time, which might be why this seems to work.

Changes:
* add 'from contextlib import contextmanager' to the imports
* add `@contextmanager' decorator to the `deploy_pulumi_script` function
* replace this `return` with `yield`

This might mean you have to do the same with `create_dummy_stack`, and use a `with` block at the call sites. Something to think about

File: .github/workflows/build.yml [hunk 10]
```diff
@@ -26,10 +27,10 @@ jobs:
 
     - name: Install Pulumi CLI
       uses: pulumi/setup-pulumi@v2
-    - name: Set up Python 3.11
+    - name: Set up Python 3.12
```
Comment (simonrw) [inline .github/workflows/build.yml:30]: Why was this needed?

**Review by simonrw — APPROVED** [review]
These changes look sensible, and I'm glad the functionality will be implemented.

Some nits that I recommend you tackle namely 
* printing error messages to `stderr`
* a sanity check about the strategy choosing logic
* checking the config strategy is valid
but apart from that, looks great!

File: bin/pulumilocal (hunk 2 above)
Comment (lakkeger) [reply bin/pulumilocal:30]: Currently we're doing the following:
1. if strategy is missing or invalid we display the chosen strategy,
2. a message that no valid configuration strategy has been chosen,
3. in the above message we're stating that we're falling back to the default strategy (overwrite) which is equivalent with the old behaviour
4. and setting the fallback strategy
5. continue the run with the fallback
Exiting with a warning would break any existing CIs imo.

File: bin/pulumilocal (hunk 3 above)
Comment (lakkeger) [reply bin/pulumilocal]: Fair enough.

File: bin/pulumilocal (hunk 4 above)
Comment (lakkeger) [reply bin/pulumilocal:209]: tbh I didn't feel the need to do all nice help messages as there is no way you'd see any of these, but thanks for beautify them :)

File: bin/pulumilocal (hunk 5 above)
Comment (lakkeger) [reply bin/pulumilocal:224]: Technically it's not an error more of a warning. Wanted to keep the luxury for the users that they don't need to reconfigure anything if they use the new version.
With your proposal any previous user's CI would start to fail with the new version.

File: bin/pulumilocal (hunk 8 above)
Comment (lakkeger) [reply bin/pulumilocal]: `overwrite` doesn't need extra steps like using a different config file (`override`) or creating a new stack (`separate`), hence no need to include here, it can fall through to reconfigure the original stack config.

File: tests/test_apply.py (hunk 9 above)
Comment (lakkeger) [reply tests/test_apply.py:122]: You are right in general, however this is a really specific usage here. I use this return value only once in the tests and that is a valid case with a correct behaviour since `tempfile.TemporaryDirectory` with Python 3.12 adds a `delete` argument. If that's been set to `False`, as a result the temp directory is not deleted and can be processed and checked. Not the nicest, fail-safest solution though I agree, but does the job.

File: .github/workflows/build.yml (hunk 10 above)
Comment (lakkeger) [reply .github/workflows/build.yml:30]: I use some features in the tests which are only available in 3.12

File: bin/pulumilocal (hunk 1 above)
Comment (lakkeger) [reply bin/pulumilocal]: Any suggestion how to call it?

File: bin/pulumilocal (hunk 6 above)
Comment (lakkeger) [reply bin/pulumilocal]: Clarify this a bit please, I'm not sure I get what you meant.
- without `--show-name` the `stack` sub-command only displays the help for itself
- `--non-interactive` is necessary because if there is no stack selected pulumi waits for input
- `-C` is to ensure we're in the right folder

File: bin/pulumilocal (hunk 1 above)
Comment (simonrw) [reply bin/pulumilocal]: Nope! Naming is hard. If you can't come up with anything convincing then I'm ok to leave it as it is.

File: bin/pulumilocal (hunk 2 above)
Comment (simonrw) [reply bin/pulumilocal:30]: I am glad we have the logging in place 👍 but regarding 

> Exiting with a warning would break any existing CIs imo.

this is a new flag and we are defaulting to the existing strategy if not set, so I don't think this is the case.

File: bin/pulumilocal (hunk 5 above)
Comment (simonrw) [reply bin/pulumilocal:224]: As per above, if I understand the change correctly, if users don't specify a `CONFIG_STRATEGY` then it will fall back to the current behaviour, so this would not break if we errored.

It is personal preference, but if I configure a tool incorrectly, I would prefer to know about it when I run it, not for it to fall back to a default with a log message which I may miss. I don't read, it turns out... 😂

File: bin/pulumilocal (hunk 6 above)
Comment (simonrw) [reply bin/pulumilocal]: I mean specifically the line I am commenting on: the 
```python
env=os.environ
```

is the default behaviour.

File: tests/test_apply.py (hunk 9 above)
Comment (simonrw) [reply tests/test_apply.py:122]: The way you have it currently you are setting `cleanup=False` in `create_dummy_stack` which as you say is not cleaned up, so the test will execute fine, however that leaves the temporary directory around. As you say this leaves the directory around after the test if someone wants to inspect it, but I feel the better behaviour is to cleanup the dir after the test, and the developer can change the `cleanup` flag manually if they are making changes. Or add a testing configuration envar to control the behaviour.

This is a fairly minor point so I'll leave the final call up to you.

File: bin/pulumilocal (hunk 6 above)
Comment (lakkeger) [reply bin/pulumilocal]: Ah, you are right. I have overlooked this in the docs.

>...these are used instead of the default behavior of inheriting the current process’ environment. ....

Apparently I'm not reading either. 😅

File: tests/test_apply.py (hunk 9 above)
Comment (lakkeger) [reply tests/test_apply.py:122]: I'm doing a not so elegant `rmtree` at the end of the test, however now that you've raised your concern perhaps this would be better done even if the test fails 🙃 Additionally not an elegant solution as you have pointed it out, perhaps this is something, among others, that with an upcoming major change like the awsx (aws-native package) support we can fix by refactoring the whole codebase.

Comment (lakkeger) [reply tests/test_apply.py:122]: As an interim solution I add a `try...finally` to the test in question, all other cases are handled with `with`.

File: bin/pulumilocal (hunk 2 above)
Comment (lakkeger) [reply bin/pulumilocal:30]: As we've agreed the fall back behaviour is only temporary until the next release.
